src/main/java/com/couchbase/lite/support/action/Action.java

```diff
@@ -0,0 +1,342 @@
+     * The `failedStep` property is set to the index of the failed perform block.
+     * @throws ActionException
+     */
+    public void run() throws ActionException {
+        try {
+            perform();
+            try {
+                cleanup(); // Ignore exception
+            } catch (ActionException e) {}
+            lastError = null;
+        } catch (ActionException e) {
+            // (perform: has already backed out whatever it did)
+            lastError = e;
+            throw e;
+        }
+    }
+
+    @Override
+    public void perform() throws ActionException {
+        if (nextStep > 0)
+            throw new IllegalStateException("Actions have already been run");
+
+        failedStep = -1;
+        for (; nextStep < peforms.size(); ++nextStep) {
+            try {
+                doAction(peforms);
+            } catch (ActionException e) {
+                failedStep = nextStep;
+                if (nextStep > 0) {
+                    try {
+                        backout(); // Ignore error
+                    } catch (ActionException backoutEx) {}
+                }
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public void backout() throws ActionException {
+        if (nextStep == 0)
+            throw new IllegalStateException("Actions have not been run");
+
+        while(nextStep-- > 0) {
+            try {
+                doAction(backouts);
+            } catch (ActionException e) {
+                Log.e(Log.TAG_ACTION, "Error backing out step: " + nextStep, e);
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public void cleanup() throws ActionException {
+        if (nextStep != peforms.size())
+            throw new IllegalStateException("Actions did not all run");
+
+        while(nextStep-- > 0) {
+            try {
+                doAction(cleanUps);
+            } catch (ActionException e) {
+                Log.e(Log.TAG_ACTION, "Error cleaning up step: " + nextStep, e);
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Subroutine that calls an action block from either performs, backOuts or cleanUps.
+     * @param actions
+     * @throws ActionException
+     */
+    private void doAction(List<ActionBlock> actions) throws ActionException {
+        try {
+            actions.get(nextStep).execute();
+        } catch (ActionException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ActionException("Exception raised by step: " + nextStep, e);
+        }
+    }
+
+    /** File-based actions: **/
+
+    /**
+     * Deletes the file/directory at the given path, if it exists.
+     * @param path Path to the file or directory to be deleted
+     * @param tempDir Path to a tem directory used for backing out the deleted file
+     * @return A delete file Action
+     */
+    public static Action deleteFile(final String path, final String tempDir) {
```

Comment:
`deleteFile()` and `moveFileToEmptyPath()` methods are an utility method. I recommend to extract these method from Action.java, and create new Java Class.